### PR TITLE
Refacted ENB config

### LINKED ISSUE #1127: Refact ENB config
We need to simplify the `.enb/make.js` file.

Part of the logic can be move by tasks:
- `.enb/tasks/dist`
- `.enb/tasks/specs`
- `.enb/tasks/tmpl-specs`
- `.enb/tasks/tests`
- `.enb/tasks/examples`
- `.enb/tasks/docs`

And also collect common code:
- `.enb/techs.js`
- `.enb/config.js`
- `.enb/helpers.js`

### PR DESCRIPTION
**Pull request by blond**
Resolved #1127

### CODE REVIEW

File: .enb/config/levels.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        desktop : ['common.blocks', 'desktop.blocks'],
+        touch : ['common.blocks', 'touch.blocks']
+    },
+    SPEC_LEVEL = {
+        path : path.join('libs', 'bem-pr', 'spec.blocks'),
+        check : false
+    };
+
+/**
+ * Returns list of leves for specified platform.
```
Comment (tadatuta) [inline .enb/config/levels.js]: leves -> levels

File: .enb/config/index.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+    LIB_NAME : 'bem-core',
```
Comment (tadatuta) [inline .enb/config/index.js]: are you sure about caps for field names?

File: .enb/helpers/page.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
+var fs = require('fs'),
+    path = require('path'),
+    config = require('../config'),
+    techs = require('../techs');
+
+/**
+ * Configures dir to build page.
+ *
+ * Before build in file system must exist directory with BEMJSON file.
```
Comment (tadatuta) [inline .enb/helpers/page.js]: Directory with BEMJSON file must exist in file system before build.

File: .enb/helpers/page.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+ *
+ * page/
+ * └── page.bemjson.js
+ *
+ * Important: BEMJSON file should be provided.
+ *
+ * After build we get js, CSS and HTML files.
```
Comment (tadatuta) [inline .enb/helpers/page.js]: js -> JS

File: .enb/helpers/page.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+ *     });
+ * };
+ */
+module.exports = function(node, opts) {
+    var nodeUrl = node.getPath();
+
+    opts = opts || {};
+
+    if(!opts.platform) {
+        var error = new Error('platform to configure the `' + nodeUrl + '` node is not specified.');
```
Comment (tadatuta) [inline .enb/helpers/page.js:52]: platform -> Platform

File: .enb/helpers/page.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+ * page/
+ * ├── page.bemjson.js
+ * ├── page.css
+ * ├── page.js
+ * └── page.html
+ *
+ * @param {NodeConfig} node — instance for configure dir
+ * @param {Object} opts
+ * @param {Object} opts.platform — platform name (desktop or touch).
+ * @param {String[]} [opts.langs] — languages.
```
Comment (tadatuta) [inline .enb/helpers/page.js]: Add `opts.bemtree` jsdoc

File: .enb/helpers/page.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        levels = [].concat(
+            config.levels(opts.platform),
+            sublevels
+        ),
+        needBEMTREE = opts.bemtree;
+
+    node.addTechs([
+        // get FileList
+        [techs.bem.levels, { levels : levels }],
+        !needBEMTREE && [techs.bem.bemjsonToBemdecl],
```
Comment (tadatuta) [inline .enb/helpers/page.js:75]: will `node.addTechs([false])` be a valid call?

File: .enb/helpers/page.js [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        // build CSS
+        [techs.css],
+
+        // build BEMHTML for browsers
+        [techs.bem.depsByTechToBemdecl, {
+            target : '?.bemhtml.bemdecl.js',
+            sourceTech : 'js',
+            destTech : 'bemhtml'
+        }],
+        [techs.bem.depsOld, {
```
Comment (tadatuta) [inline .enb/helpers/page.js:88]: any reason for `depsOld` in `bem-core`?

File: .enb/helpers/page.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        // build HTML using BEMTREE
+        node.addTechs([
+            [techs.engines.bemhtml],
+            [techs.engines.bemtree],
+            [techs.html.bemtree]
+        ]);
+    } else {
+        node.addTechs([
+            // build HTML using BEMHTML
+            [techs.engines.bemhtml],
```
Comment (tadatuta) [inline .enb/helpers/page.js]: why not to add `[techs.engines.bemhtml]` outside `if`?

File: .enb/helpers/page.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+            // build HTML using BH
+            [techs.engines.bhCommonJS, {
+                devMode : false,
+                bhOptions : {
+                    jsAttrName : 'data-bem',
+                    jsAttrScheme : 'json'
+                }
+            }],
+            [techs.html.bemhtml, { target : '?.bh.html' }]
```
Comment (tadatuta) [inline .enb/helpers/page.js]: is `html.bemhtml` ok for `?.bh.html`?

File: .enb/helpers/page.js [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            [techs.html.bemtree]
+        ]);
+    } else {
+        node.addTechs([
+            // build HTML using BEMHTML
+            [techs.engines.bemhtml],
+            [techs.html.bemhtml],
+
+            // build HTML using BH
+            [techs.engines.bhCommonJS, {
```
Comment (tadatuta) [inline .enb/helpers/page.js:133]: do we really always need bemhtml and bh at the same time?

File: .enb/helpers/page.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+                    jsAttrScheme : 'json'
+                }
+            }],
+            [techs.html.bemhtml, { target : '?.bh.html' }]
+        ]);
+
+        node.addTarget('?.bh.html');
+    }
+
+    node.addTargets([
```
Comment (tadatuta) [inline .enb/helpers/page.js:146]: can't we switch to `min.{js,css}` here? are there any drawbacks?

File: .enb/helpers/page.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        '_?.css', '_?.js',
+        '?.html'
+    ]);
+
+    node.mode('development', function() {
+        node.addTechs([
+            [techs.files.copy, { source : '?.css', target : '_?.css' }],
```
Comment (tadatuta) [inline .enb/helpers/page.js:153]: isn't symlink tech better here?

File: .enb/helpers/page.js [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+            [techs.files.copy, { source : '?.js', target : '_?.js' }]
+        ]);
+    });
+
+    node.mode('production', function() {
+        node.addTechs([
+            [techs.borschik, { source : '?.css', target : '_?.css', freeze : true, tech : 'cleancss' }],
```
Comment (tadatuta) [inline .enb/helpers/page.js]: isn't `freeze : true` a default value now?

File: .enb/helpers/page.js [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+            [techs.borschik, { source : '?.js', target : '_?.js', freeze : true }]
+        ]);
+    });
+
+    langs.forEach(function(lang) {
```
Comment (tadatuta) [inline .enb/helpers/page.js:165]: why is it just coping?

File: .enb/make.js [hunk 16]
```diff
@@ -1,347 +1,47 @@
...
 
-    config.setLanguages(langs? langs.split(' ') : [].concat(DEFAULT_LANGS));
+    // load task configs
+    [
+        'dist',
+        'specs', 'tmpl-specs',
+        'tests',
+        'docs', 'examples',
+    ].forEach(function (name) {
+        var url = [__dirname, 'tasks', name + '.js'].join('/');
```
Comment (tadatuta) [inline .enb/make.js]: shouldn't it be `path.join()`?

File: .enb/tasks/dist.js [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+            sources : ['.tmp.source.js', LIB_NAME + '.dev.bh.js']
+        }],
+        [techs.ym, {
+            source : '.tmp.source+bh.js',
+            target : LIB_NAME + '.dev.js+bh.js'
+        }],
+        [techs.borschik, { source : LIB_NAME + '.dev.js+bh.js', target : LIB_NAME + '.js+bh.js' }]
+    ]);
+
+    node.addTargets([
```
Comment (tadatuta) [inline .enb/tasks/dist.js:101]: what do you think about

``` js
['dev.css', 'dev.js', ...].map(function(ext) {
    return LIB_NAME + '.' + ext;
});
```

?

File: .enb/tasks/docs.js [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Creates `docs` task.
+ *
+ * This task allows you to build docs.
+ *
+ * @param {ProjectConfig} project - main ENB config for this project
+ * @example Build all docs
+ * $ magic run specs
+ * @example Build docs for desktop platfrom
```
Comment (tadatuta) [inline .enb/tasks/docs.js]: platfrom -> platform

File: .enb/config/index.js (hunk 2 above)
Comment (blond) [inline .enb/config/index.js]: no )

File: .enb/tasks/docs.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+module.exports = function (project) {
+    // load plugin
+    project.includeConfig('enb-bem-docs');
+    var plugin = project.module('enb-bem-docs'),
+        // create task with `specs` name
+        // and get helper to configure it
+        helper = plugin.createConfigurator('docs');
+
+    PLATFORMS.forEach(function (platform) {
+        configure(helper, platform, project.getLanguages());
```
Comment (tadatuta) [inline .enb/tasks/docs.js]: shouldn't we cache `project.getLanguages()` outside the loop?

File: .enb/tasks/examples.js [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+ * @example Build examples for desktop platform
+ * $ magic make desktop.examples
+ */
+module.exports = function (project) {
+    // load plugin
+    if(!project._modules['enb-bem-examples']) {
+       project.includeConfig('enb-bem-examples');
+    }
+    var plugin = project.module('enb-bem-examples'),
+        // create task with `specs` name
```
Comment (tadatuta) [inline .enb/tasks/examples.js]: why `specs`?

File: .enb/helpers/page.js (hunk 5 above)
Comment (blond) [inline .enb/helpers/page.js:52]: Why?

```
ConfigError: platform to configure the `desktop.bundles/index` node is not specified.
```

File: .enb/helpers/page.js (hunk 7 above)
Comment (blond) [inline .enb/helpers/page.js:75]: no, below there is a filter

File: .enb/tasks/specs.js [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+function configure(helper, platform) {
+    var dir = platform + '.specs';
+
+    helper.configure({
+        destPath     : dir,
+        levels       : config.levels(platform),
+        sourceLevels : config.levels(platform, { specs : true }),
+        jsSuffixes   : ['vanilla.js', 'browser.js', 'js'],
+        specSuffixes : ['spec.js'],
+        depsTech     : techs.bem.depsOld
```
Comment (tadatuta) [inline .enb/tasks/specs.js:44]: any reason to use depsOld?

File: .enb/helpers/page.js (hunk 8 above)
Comment (blond) [inline .enb/helpers/page.js:88]: No, but this PR about refactoring only.

All improvements suggest doing separately.

File: .enb/helpers/page.js (hunk 11 above)
Comment (blond) [inline .enb/helpers/page.js:133]: No, but this PR about refactoring only.

All improvements suggest doing separately.

File: .enb/tasks/tests.js [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+    // load plugin
+    if(!project._modules['enb-bem-examples']) {
+       project.includeConfig('enb-bem-examples');
+    }
+    var plugin = project.module('enb-bem-examples'),
+        // create task with `specs` name
+        // and get helper to configure it
+        helper = plugin.createConfigurator('tests');
+
+    PLATFORMS.forEach(function (platform) {
```
Comment (tadatuta) [inline .enb/tasks/tests.js:27]: isn't it possible to move lines 26-40 into separate module and reuse all over?

File: .enb/helpers/page.js (hunk 12 above)
Comment (blond) [inline .enb/helpers/page.js:146]: This PR about refactoring only.

All improvements suggest doing separately.

File: .enb/helpers/page.js (hunk 13 above)
Comment (blond) [inline .enb/helpers/page.js:153]: may be

File: .enb/helpers/page.js (hunk 5 above)
Comment (tadatuta) [inline .enb/helpers/page.js:52]: oh, ok

File: .enb/helpers/page.js (hunk 8 above)
Comment (tadatuta) [inline .enb/helpers/page.js:88]: ok

File: .enb/helpers/page.js (hunk 11 above)
Comment (tadatuta) [inline .enb/helpers/page.js:133]: will you create tasks for them?

File: .enb/helpers/page.js (hunk 15 above)
Comment (blond) [inline .enb/helpers/page.js:165]: because in `bem-core` there are no files for i18n.

File: .enb/tasks/dist.js (hunk 17 above)
Comment (blond) [inline .enb/tasks/dist.js:101]: fixed

File: .enb/tasks/tests.js (hunk 22 above)
Comment (blond) [inline .enb/tasks/tests.js:27]: I think that this is unnecessary complication

File: .enb/tasks/specs.js (hunk 21 above)
Comment (blond) [inline .enb/tasks/specs.js:44]: This PR about refactoring only.

All improvements suggest doing separately.

File: .enb/helpers/page.js (hunk 8 above)
Comment (blond) [inline .enb/helpers/page.js:88]: https://github.com/bem/bem-core/issues/1131

File: .enb/helpers/page.js (hunk 13 above)
Comment (blond) [inline .enb/helpers/page.js:153]: https://github.com/bem/bem-core/issues/1132

File: .enb/helpers/page.js (hunk 11 above)
Comment (blond) [inline .enb/helpers/page.js:133]: https://github.com/bem/bem-core/issues/1133

File: .enb/helpers/page.js (hunk 12 above)
Comment (blond) [inline .enb/helpers/page.js:146]: https://github.com/bem/bem-core/issues/1128

**Comment by blond** [comment]
:up:

**Pull request merged**